pkg/trace: implement support for OpenTelemetry collecting

gbbr opened the pull request:
This change adds a new OTLP receiver in the agent which is off by
default. It can be enabled by setting the following config (also by env.
vars):

    experimental:
      otlp:
        grpc_port: 50052 # enables gRPC receiver
        http_port: 50051 # enables HTTP receiver

Environment variables are:

- DD_OTLP_HTTP_PORT
- DD_OTLP_GRPC_PORT

gbbr committed:
pkg/trace/pb: add otlppb package

This change copies over all necessary trace & collector protobufs from:

     • https://github.com/open-telemetry/opentelemetry-collector/tree/v0.26.0/internal/data/protogen
     • https://github.com/open-telemetry/opentelemetry-proto/tree/v0.9.0/opentelemetry/proto

gbbr committed:
pkg/trace/{config,api,agent}: add and enable OTLP receiver

This change adds a new OTLP receiver in the agent which is off by
default. It can be enabled by setting the following config (also by env.
vars):

    apm_config:
      otlp:
        grpc_port: 50052 # enables gRPC receiver
        http_port: 50051 # enables HTTP receiver

Environment variables are:

- DD_APM_OTLP_HTTP_PORT
- DD_APM_OTLP_GRPC_PORT

gbbr committed:
pkg/trace/api: add telemetry to OTLPReceiver

gbbr committed:
pkg/trace/api: add more OTLP* unit tests

gbbr committed:
releasenotes: add release note

ericmustin commented on pkg/trace/api/api.go:
```diff
@@ -724,6 +725,9 @@ func tracesFromSpans(spans []pb.Span) pb.Traces {
 // is empty or no tags are found, an empty string is returned.
 func getContainerTags(containerID string) string {
+	if containerID == "" {
+		return ""
```
these container tags should be available as resource attributes in opentelemetry, i will try to investigate what attributes to use here, but I don't think this is blocking.

gbbr replied:
The container tags are obtained from the orchestrator API (k8s, ECS, etc). I wouldn't count on OTLP clients to send it because I doubt apps want to query it and add latency on the client side. I'd advise we keep these because they are very useful and key to the Datadog product.

ericmustin reviewed (changes requested):
A few small questions and feedback but overall this looks great! I will try to work out open questions around sampling this week, but nice work on this! 🎉

gbbr commented on pkg/trace/config/testdata/full.yaml:
```diff
@@ -26,6 +26,9 @@ apm_config:
       - "apikey5\n \n         "
   env: test
+  otlp:
+    http_port: 50051
+    grpc_port: 50052
```
That's no problem. This is just a test case. Users can chose whatever ports they like. When we document these features, we can use the ports you are proposing in the official documentation.

gbbr commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,426 @@
+	o.processRequest(otlpProtocolHTTP, req.Header, &in)
+}
+
+func tagsFromHeaders(h http.Header, protocol string) []string {
```
This is metadata the agent needs for internal metrics. In my OTLP usage example (check [this](https://gist.github.com/gbbr/4a54dd02d34ad05e694952e0a02e1c67#file-otlp_datadog-go-L31-L35)) I send it as headers, just like we do for our tracers. If you think there would be a better way, I'm open to change the approach in the client and change the agent. We should figure out what is most user-friendly.

It's not a must, but it's nice to have...

gbbr commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,426 @@
+	name := spanKindName(in.Kind)
+	if lib.Name != "" {
+		name = lib.Name + "." + name
+	}
```
Done.

gbbr commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,426 @@
+	if lib.Version != "" {
+		span.Meta["instrumentation_library.version"] = lib.Version
+	}
+	if span.Service == "" && span.Meta["peer.service"] != "" {
```
Done.

gbbr commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,426 @@
+	"github.com/DataDog/datadog-agent/pkg/trace/pb/otlppb"
+	"github.com/DataDog/datadog-agent/pkg/trace/sampler"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
```
I'd prefer to not import anything from OTLP but if we ever reach a point where this is a must, we can.

gbbr commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,426 @@
+				span.Meta["error.msg"] = anyValueString(attr.Value)
+			case "exception.type":
+				span.Meta["error.type"] = anyValueString(attr.Value)
+			case "exception.stack":
```
Done!

gbbr committed:
Addressed comments

gbbr reviewed (commented):
@ericmustin thanks for the review, please check the latest commit. I hope I haven't missed anything!

ericmustin reviewed (approved):
I think this looks good to me, we still have open questions on 

- any sampling and trace-stats limitations and recommend sdk/client configuration
- clearly specify which of the 4 export formats we support in the compliance matrix( here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#exporters)  
- determine if there is enhaned metadata related to containers, fargate, or instrumentation library information (like application language, tracer version, etc) that we can reliably grab as well.

but i don't believe any of that is blocking here and I can provide that async. Perhaps we may want to explicitly mark this experimental until we have more robust user feedback though on live data.

Nice work @gbbr !

seanyu4296 commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,428 @@
+	if len(b) < 8 {
+		return 0
+	}
+	return binary.LittleEndian.Uint64(b)
```
just wondering is this the same implementation as the one here?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/f1e8331be111a4a8bf65670dfaff359714edef85/exporter/datadogexporter/translate_traces.go#L465

it looks different but im not a pro in golang. might be best to align 

https://docs.datadoghq.com/tracing/connect_logs_and_traces/opentelemetry/

gbbr replied:
Not entirely the same, I think. I was wondering about this myself, thanks for pointing me to the link. I will update this.

gbbr replied:
I ran some tests ([link](https://play.golang.org/p/Wm9yZagS54z)) and updated the algorithm to match the output, in the latest commit. PTAL

seanyu4296 replied:
I'm not an entirely sure of this I'm not an employee of Datadog 😅 maybe @ericmustin can confirm

gbbr replied:
It's fine. It's the same thing. Thanks for the comment.

gbbr committed:
fix ID conversion

gbbr committed:
pkg/trace/api: add a benchmark and some micro-improvements

gbbr committed:
pkg/trace/api: avoid json.Marshal

```
name              old time/op    new time/op    delta
ProcessRequest-8    11.8µs ± 0%     7.8µs ± 0%  -34.12%

name              old alloc/op   new alloc/op   delta
ProcessRequest-8    6.64kB ± 0%    5.75kB ± 0%  -13.43%

name              old allocs/op  new allocs/op  delta
ProcessRequest-8      64.0 ± 0%      68.0 ± 0%   +6.25%
```

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -360,13 +360,13 @@ const (
 )
 
-func (r *HTTPReceiver) tagStats(v Version, req *http.Request) *info.TagStats {
+func (r *HTTPReceiver) tagStats(v Version, header http.Header) *info.TagStats {
```
I guess it's a leftover from a previous version of the code. I could undo it if you like.

ericmustin commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,494 @@
+	metrics.Count("datadog.trace_agent.otlp.bytes", int64(len(slurp)), mtags, 1)
+	var in otlppb.ExportTraceServiceRequest
+	switch getMediaType(req) {
+	case "application/x-protobuf":
```
I just want to confirm that we are supporting here , per the compliance matrix: (here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#exporters) 

`OTLP/HTTP binary Protobuf Exporter` format. Is that correct?

ericmustin commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,494 @@
+	case "application/json":
+		fallthrough
+	default:
+		if err := json.Unmarshal(slurp, &in); err != nil {
```
Similar to above comment, i just want to confirm that this is support of (per compliance matrix: https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#exporters) 

`OTLP/HTTP JSON Protobuf Exporter` format. Is that correct?

gbbr replied:
Yes, to both questions. And seeing that table now, I added a new commit to support gzip too, so we can tick that box.

gbbr committed:
Add support for gzip

mx-psi commented on pkg/trace/config/testdata/full.yaml:
```diff
@@ -26,6 +26,9 @@ apm_config:
       - "apikey4              "
       - "apikey5\n \n         "
   env: test
+  otlp:
+    http_port: 50051
+    grpc_port: 50052
```
Is it worth making more explicit here that this is experimental?
```suggestion
  experimental:
    otlp:
      http_port: 50051
      grpc_port: 50052
```

Also, revisiting this, if we want to add more features that are HTTP or GRPC specific it will be easier if it's
```
http:
  port: 50051
```
instead of `http_port`

gbbr replied:
I'll address both your points:
1. The problem is that there's no going back from that config once we add it (`experimental`). If we really want to emphasise that, I'd rather add an extra layer on top such as feature flag.
2. On your second point. I'm on the fence. I tend to try and keep things simple. I can't think of something we might need, and maybe we'll never need anything, so I don't want to pre-maturely prepare for something that might not happen. Worst case scenario, we can always add the `http` group in parallel with `http_port` as a section for configuring HTTP behaviour and it wouldn't be that bad. I see it a little worse to add extra indentation and sections prematurely.

Open to discuss this further if you feel otherwise.

mx-psi replied:
> I'd rather add an extra layer on top such as feature flag. We can also mention this in documentation
Both sound like good ideas to me

> I can't think of something we might need, and maybe we'll never need anything, so I don't want to pre-maturely prepare for something that might not happen

It may help to have a look at the OTLP receiver settings on the OpenTelemetry Collector https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/config.md we may not want to support most of those settings ever but if it's possible that we will want to support some of them in the future I think it's a good argument to use `http`/`grpc` groups

gbbr replied:
I've moved these under experimental, and left them like this. We can always change them now that we've concluded that experimental is bound to change and the user is aware of that.

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,506 @@
+		}
+	}
+	if _, ok := span.Meta["env"]; !ok {
+		if env := span.Meta["deployment.environment"]; env != "" {
```
We should not use raw strings: semantic conventions are still unstable and it's better to import them from a package like https://pkg.go.dev/go.opentelemetry.io/otel@v0.20.0/semconv. The package might change in the future (see open-telemetry/opentelemetry-collector#2600) so we could do it later.

Another issue here is that at some point OpenTelemetry will introduce telemetry schemas (see doc [here](https://github.com/open-telemetry/oteps/blob/main/text/0152-telemetry-schemas.md)) which may change these constants. It's still early so I don't think this is a concern for now but it's something to keep in mind for designing this.

gbbr replied:
I tried to avoid importing any OpenTelemetry stuff, but I guess this is a strong point. We should probably change this...

gbbr replied:
Done.

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,506 @@
+}
+
+// spanKind2Type returns a span's type based on the given kind and other present properties.
+func spanKind2Type(kind otlppb.Span_SpanKind, span *pb.Span) string {
```
These (and other bits) are basically the same code we have on the OpenTelemetry Collector Datadog exporter. Do we have a plan to share code between the two in the future? I want to do that for metrics so I can help with it

gbbr replied:
No plan, but probably a good idea ad something we should / could consider doing!

gbbr replied:
It will probably be a good idea. We should likely make this one of those exported datadog-agent modules.

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,506 @@
+}
+
+// anyValueString converts otlppb.AnyValue a to its string representation.
+func anyValueString(a *otlppb.AnyValue) string {
```
On the metrics side we use https://github.com/open-telemetry/opentelemetry-collector/blob/3f5c7180c51ed67a6f54158ede5e523822e9659e/translator/trace/protospan_translation.go#L61 to make this translation and optionally add them as metric tags (the types are different because of the `pdata` package but it's the same thing). If this affects discoverability we should try and do the same translation.

gbbr replied:
Not entirely sure what you're proposing, but I feel its best we keep this in our own hands because its on the hot path (runs on each span multiple times) and we want to get this as performant as possible. Furthermore the code you linked to can be quite slow if it has to marshal JSON using the `reflect` package...

mx-psi replied:
So, my concern is if the Datadog webapp uses these span metadata fields to correlate with Datadog metric tags.

Imagine we have a field `key` with value `0.9`. With your code this gets mapped to `"0.90"`, while with the OpenTelemetry one it gets mapped to `"0.9"` (see https://play.golang.org/p/O4q9xHQcguU). That means we would get a `key:0.9` tag on metrics and a `key:0.90` field on the trace metadata. If the Datadog webapp uses this to correlate the metric and the trace (not sure if it does) then it won't find one from the other since they have different strings as values. The same thing happens with other fields.

gbbr replied:
I would leave it for now if it's ok with you, because this is definitely hot path and gets called a lot. The linked implementation uses `json.Marshal` which is very slow, while this one is optimised. If we run into issues, we can always change it (experimental).

gbbr committed:
pkg/trace/config: put otlp under experimental

gbbr committed:
pkg/trace/api: user go.opentelemetry.io/otel/semconv

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,507 @@
+		}
+	} else if m := meta["messaging.operation"]; m != "" {
+		r = m
+		if dest := meta["messaging.destination"]; dest != "" {
```
```suggestion
		if dest := meta[string(semconv.MessagingDestinationKey)]; dest != "" {
```

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,507 @@
+		typ = "web"
+	case otlppb.Span_SPAN_KIND_CLIENT:
+		typ = "http"
+		db, ok := span.Meta["db.system"]
```
```suggestion
		db, ok := span.Meta[string(semconv.DBSystemKey)]
```

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,507 @@
+				LangVersion:     fastHeaderGet(header, headerLangVersion),
+				Interpreter:     fastHeaderGet(header, headerLangInterpreter),
+				LangVendor:      fastHeaderGet(header, headerLangInterpreterVendor),
+				TracerVersion:   fmt.Sprintf("otlp-%s", rattr["telemetry.sdk.version"]),
```
```suggestion
				TracerVersion:   fmt.Sprintf("otlp-%s", rattr[string(semconv.TelemetrySDKVersionKey)]),
```

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,507 @@
+		for _, attr := range rspans.Resource.Attributes {
+			rattr[attr.Key] = anyValueString(attr.Value)
+		}
+		lang := rattr["telemetry.sdk.language"]
```
```suggestion
		lang := rattr[string(semconv.TelemetrySDKLanguageKey)]
```

mx-psi commented on pkg/trace/api/otlp.go:
```diff
@@ -0,0 +1,507 @@
+		}
+		for _, attr := range e.Attributes {
+			switch attr.Key {
+			case "exception.message":
```
Here you can use `ExceptionMessageKey`, `ExceptionTypeKey` and `ExceptionStacktraceKey`

mx-psi reviewed (approved):
Approving since what remains (using a bunch of constants from the `semconv` package) can be done on a different PR. I left the comments anyway but feel free to address them on a separate PR!

gbbr committed:
Add semconv.Exception{...}Key keys

gbbr committed:
pkg/trace/api: user more semconv stuff

gbbr committed:
license

mx-psi reviewed (approved):
LGTM

remeh reviewed (approved):
Note that you've mentioned a different config parameter in the changelog than the one you declare in the `pkg/trace/config`.

gbbr committed:
correct release note
